lib/Frontend/Frontend.cpp

```diff
@@ -474,12 +474,29 @@ void CompilerInstance::setupOutputBackend() {
 
   // Mirror the output into CAS.
   if (supportCaching()) {
+    auto &InAndOuts = Invocation.getFrontendOptions().InputsAndOutputs;
     CASOutputBackend = createSwiftCachingOutputBackend(
-        *CAS, *ResultCache, *CompileJobBaseKey,
-        Invocation.getFrontendOptions().InputsAndOutputs,
+        *CAS, *ResultCache, *CompileJobBaseKey, InAndOuts,
         Invocation.getFrontendOptions().RequestedAction);
+
+    auto OutputFiles = InAndOuts.copyOutputFilenames();
+    std::unordered_set<std::string> OutputFileSet(
+        std::make_move_iterator(OutputFiles.begin()),
+        std::make_move_iterator(OutputFiles.end()));
+    // Filter the object file output if MCCAS is enabled, we do not want to
+    // store the object file itself, but store the MCCAS CASID instead.
+    auto FilterBackend = llvm::vfs::makeFilteringOutputBackend(
+        CASOutputBackend,
+        [&, OutputFileSet](StringRef Path,
+                           std::optional<llvm::vfs::OutputConfig> Config) {
+          if (InAndOuts.getPrincipalOutputType() != file_types::ID::TY_Object)
+            return true;
+          return !(Invocation.getIRGenOptions().UseCASBackend &&
```

Comment:
I would rather only create the filtering backend when `UseCASBackend`, otherwise just don't create it.